File: javascript/programmers/javascript/level3/섬연결하기.js
const n = 4;
const costs = [
  [0, 1, 1],
  [0, 2, 2],
  [2, 3, 1],
];

const solution = (n, costs) => {
  let answer = 0;
  const isVisited = new Array(n).fill(false);
  const isBridge = new Array(costs.length).fill(false);
  costs.sort((a, b) => {
    return a[2] - b[2];
  });
  let num = 0;

  isVisited[costs[0][0]] = true;
  isVisited[costs[0][1]] = true;
  isBridge[0] = true;
  answer += costs[0][2];
  num += 1;

  while (num < n - 1) {
    for (let i = 1; i < costs.length; i++) {
      const [start, end, cost] = costs[i];
      if (
        !isBridge[i] &&
        ((!isVisited[start] && isVisited[end]) ||
          (isVisited[start] && !isVisited[end]))
      ) {
        num++;
        answer += cost;
        isVisited[start] = true;
        isVisited[end] = true;
        isBridge[i] = true;
        break;
      }
      console.log(isVisited);
    }
  }
  return answer;
};

console.log(solution(n, costs));

File: python/programmers/level3/섬연결하기.py
def solution(n, costs):
    answer = 0

    def find(target):
        if target == parent[target]:
            return target

        parent[target] = find(parent[target])
        return parent[target]

    def union(a, b):
        a = find(a)
        b = find(b)

        if a < b:
            parent[b] = a
        else:
            parent[a] = b

    parent = [i for i in range(n)]
    costs.sort(key=lambda x: x[2])

    edgeCount = 0
    for cost in costs:
        a, b, weight = cost
        if find(a) != find(b):
            union(a, b)
            answer += weight
            edgeCount += 1
        if edgeCount == n - 1:
            break

    return answer


print(solution(4, [[0, 1, 1], [0, 2, 2], [1, 2, 5], [1, 3, 1], [2, 3, 8]]))
